scripts/release.js

```diff
@@ -0,0 +1,120 @@
+;(async () => {
+  const fs = require('fs')
+  const { execSync } = require('child_process')
+  const { Octokit } = require('@octokit/rest')
+
+  const [, , action, fileName] = process.argv
+
+  const actions = {
+    check: async () => {
+      const {
+        GITHUB_REPOSITORY,
+        GH_TOKEN,
+        GITHUB_SHA,
+        GITHUB_RUN_ID,
+      } = process.env
+
+      const { owner, repo } =
+        /^(?<owner>[^/]+)\/(?<repo>.*)$/.exec(GITHUB_REPOSITORY)?.groups || {}
+
+      const gh = new Octokit({ auth: GH_TOKEN })
+
+      const searchParams = {
+        is: 'pr',
+        repo: GITHUB_REPOSITORY,
+        sha: GITHUB_SHA,
+      }
+
+      const searchQuery = Object.entries(searchParams)
+        .map(([key, value]) => `${key}:${value}`)
+        .join(' ')
+
+      const {
+        data: { items: prs },
+      } = await gh.search.issuesAndPullRequests({
+        q: searchQuery,
+      })
+
+      const skipReleaseLabel = 'skip release'
+      if (
+        // if any PRs attributed to this SHA have a `skip release` label
+        prs.some(({ labels }) =>
+          labels.find(({ name }) => name === skipReleaseLabel)
+        )
+      ) {
+        console.log(
+          `Identified matching PR with "${skipReleaseLabel}" label. Cancelling deployment.`
+        )
+        try {
+          await gh.actions.cancelWorkflowRun({
+            owner,
+            repo,
+            run_id: GITHUB_RUN_ID,
+          })
+
+          // Block the workflow permanently until it is shut down by GitHub
+          await new Promise(() => null)
```

Comment:
did you do any research that led you to this approach? it seems to be relying on an assumption that GH will continue behaving a certain way

Reply:
> Note that cancellation is not immediate (takes about 15 seconds), so I added a sleep in the case we want to cancel to not mark any step as “failed”. It’s mostly a UI thing because the whole run is reported as cancelled even without the sleep.
> ([Source](https://github.community/t/is-there-a-way-to-cancel-a-running-workflow-within-one-of-its-jobs/17493/4))

Reading the above, I wasn't sure if GitHub would return _after_ the workflow had been stopped, or after the task to stop the workflow was queued. If it was the later scenario, this workflow could possibly be able to continue for 15 or so seconds until GitHub quashes it. The `await new Promise(() => null)` is my way of preventing this step from finishing so that if that scenario is true, subsequent steps do not start running.